AppController/djinn.rb

```diff
@@ -5240,62 +5242,64 @@ def scale_down_instances
           # Right now, only the autoscaled machines are started with just the
           # compute role, so we check specifically for that during downscaling
           # to make sure we only downscale the new machines added.
-          node_to_remove = nil
           if node.jobs == ['compute']
             Djinn.log_info("Removing node #{node}")
-            node_to_remove = node
+            nodes_to_remove << node
           end
-
-          num_terminated = terminate_node_from_deployment(node_to_remove)
-          num_scaled_down += num_terminated
         }
+        terminate_nodes_from_deployment(nodes_to_remove)
       }
 
       # Make sure we have the proper list of blobservers configured.
       update_blob_servers
     }
   end
 
-  # Removes the specified node from the deployment and terminates
+  # Removes the specified nodes from the deployment and terminates
   # the instance from the cloud.
   #
   # Args:
-  #   node_to_remove: A node instance, to be terminated and removed
-  #     from this deployment.
-  def terminate_node_from_deployment(node_to_remove)
-    if node_to_remove.nil?
-      Djinn.log_warn("Tried to scale down but couldn't find a node to remove.")
+  #   nodes_to_remove: A node instance, or an Array of nodes to be
+  #                    terminated and removed from this deployment.
+  def terminate_nodes_from_deployment(nodes_to_remove)
+    if nodes_to_remove.nil? || nodes_to_remove.empty?
+      Djinn.log_info("Tried to scale down but couldn't find a node to remove.")
       return 0
     end
 
+    # Let's get the list of IDs of the instances to terminate.
+    instance_ids = nodes_to_remove.map{ |node| node.instance_id }
+
     imc = InfrastructureManagerClient.new(@@secret)
     begin
-      imc.terminate_instances(@options, node_to_remove.instance_id)
+      imc.terminate_instances(@options, instance_ids)
```

Comment:
I think error handling would need to be changed to safely allow terminating multiple instances.

When terminating instances one at a time failures are handled instance by instance. When terminating a group of instances it looks like the failure semantics are agent specific (i.e. azure/gce may terminate some instances and fail) and the state of each instance would need to be verified after a termination error.